File: 프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <map>
using namespace std;

// vector<int> mbti(26);

char mbti[4][2] = {{'R','T'},{'C','F'},{'J','M'},{'A','N'}};

string solution(vector<string> survey, vector<int> choices) {
    map<char, int> score;
    
    for (int i = 0; i < survey.size(); i++)
    {
        string type = survey[i];
        
        if (choices[i] < 5)
            score[type[0]] += (4 - choices[i]);
        else
            score[type[1]] += (choices[i] % 4);
    }
    
        string answer = "";
    
        for (int i = 0; i < 4; i++){
            auto type = mbti[i];

            if (score[type[0]] < score[type[1]])
                answer += type[1];
            else
                answer += type[0];
            
        }
        cout << endl;
    
    return answer;
}

File: 프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.py
def solution(survey, choices):
    state = {i: 0 for i in "RTCFJMAN"}
    
    for i, s in enumerate(survey):
        n, p = s[0], s[1]
        score = choices[i]
        
        if (score < 4):
            state[n] += 4 - score
        else:
            state[p] += score - 4
    # print(state)
    answer = ['R','C','J','A']
    
    if state['T'] > state['R']:
        answer[0] = 'T'
    if state['F'] > state['C']:
        answer[1] = 'F'
    if state['M'] > state['J']:
        answer[2] = 'M'
    if state['N'] > state['A']:
        answer[3] = 'N'
        
    return ''.join(answer)
